Setup segment

GTcharles opened the pull request:
This PR adds segment as one of the javascript snippets (following GTM's implementation).

It also invokes segment code when users are logged and view the course page as well as when they verify their certificates.

caws committed:
Setup segment.js.
The setup follows the current implementation of GoogleTagManager and uses a SEGMENT_WRITE_KEY environment variable.

caws committed:
Require segment lib.

caws committed:
Expose segment instance.

caws committed:
Segment as a helper method.

caws committed:
Must use wirte_key here too.

caws committed:
Use #blank? as opposed to #present?

caws committed:
Add SEGMENT_WRITE_KEY env var entries.

caws committed:
Setup course viewed event.

caws committed:
Broadcast course completed event.

caws committed:
Use presenter.
The course name is only present in the presenter.

caws committed:
Add segment to content security policy.

mpraglowski commented on app/views/issued_certificates/verify.html.erb:
```diff
@@ -3,6 +3,20 @@
+  <% if current_user %>
+    <%= javascript_tag nonce: true do %>
+      analytics.track(
+      'Course Completed',
```
I'm not sure this is the valid event - `Course Completed` would be send to Segment each time user will visit certificate page. This could be send from backend code.

GTcharles replied:
That's true. This should be 'Certificate Viewed' if anything.
I'll work on this a bit more to trigger the call from ruby code instead.

caws committed:
Issue Certificate Viewed event when visiting certificate page

caws committed:
Rename Segment class to SegmentSnippet.
This is because the analytics gem also has a Class named Segment, which collides with this one.

caws committed:
Add Segment analytics gem.

caws committed:
Inject Segment::Analytics into AfterCourseCompletionService.
This will let Segment know that a user has finished their course.

caws committed:
Use logger.

caws committed:
Appease Rubocop.

caws committed:
Mock external_id for User factory.

caws committed:
Skip analytics event if no external_id is present.

GTcharles commented:
Alright, it should be okay now. I added the analytics gem to send the event when the `AfterCourseCompletionService` is called. When viwing the certificate it triggers the 'Certificate Viewed' event.

GTcharles commented:
Morning. Are we good to merge this?

gt-ch commented on app/services/students/after_course_completion_service.rb:
```diff
@@ -3,13 +3,29 @@ module Students
       student,
-      notification_service: Developers::NotificationService.new
+      notification_service: Developers::NotificationService.new,
+      analytics: Segment::Analytics.new(
```
I don't really mind because we are going away from this project.

But for elegance, we can just add it to Rails initializer as described in the Segment documentation.
Then instead of polluting base controller, just create a Concern and add it to the controllers wherever required.

gt-ch replied:
And in services or other classes then we can just use the global object `SegmentAnalytics` for example that we set up in Initializer.

GTcharles replied:
Yeah, since we were not going to be working on this anymore I figured it wasn't worth it spending too much time on the implementation.

For the other projects that are more long term I used an adapter approach where the adapter is loaded during initialization ([like this](https://github.com/growthtribeacademy/gt-assessment-app/blob/432c6d285706c7f9504c29e4734139b7a97ac0ef/config/environments/development.rb#L67). The Segment analytics class is then injected as a dependency and we can then add other analytics tools to the adapter if needed and keep using the underlying implementation. The adapter can be used from anywhere using `Rails.configuration.analytics_adapter`.

gt-ch commented on lib/segment_snippet.rb:
```diff
@@ -0,0 +1,74 @@
+    return if segment_write_key.blank?
+
+    <<-HTML
+      (function () {
```
Have you written it yourself or given in some Segment documentation?

GTcharles replied:
Nope, got it from Segment itself :sweat_smile:. It's just their JS lib really, but pupilfirst implements these JS snippets as part of the ruby code for some reason. Probably a legacy aspect from previous versions of rails.

gt-ch commented on spec/factories/user.rb:
```diff
@@ -5,6 +5,7 @@
     school { School.find_by(name: 'test') || create(:school, :current) }
     title { Faker::Lorem.words(number: 3).join(' ') }
     time_zone { ENV['SPEC_USER_TIME_ZONE'] || 'Asia/Kolkata' }
+    external_id { 'c2896708-78c1-40f1-9f17-99f132335b47' }
```
I would generate it unique for each User record instead of hard coding it. But ignore for this project now.

GTcharles replied:
Indeed, that would be better.
